pkg/virtual/syncer/coordination/helpers.go

```diff
@@ -0,0 +1,228 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package coordination
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	syncercontext "github.com/kcp-dev/kcp/pkg/virtual/syncer/context"
+	"github.com/kcp-dev/kcp/pkg/virtual/syncer/transformations"
+)
+
+func FilteredSyncerViewsChanged(old, new metav1.Object, keepSyncTarget func(syncTargetKey string) bool) bool {
+	oldSyncerViewAnnotations := make(map[string]string)
+	for name, value := range old.GetAnnotations() {
+		if strings.HasPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix) {
+			if keepSyncTarget(strings.TrimPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix)) {
+				oldSyncerViewAnnotations[name] = value
+			}
+		}
+	}
+
+	newSyncerViewAnnotations := make(map[string]string)
+	for name, value := range new.GetAnnotations() {
+		if strings.HasPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix) {
+			if keepSyncTarget(strings.TrimPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix)) {
+				newSyncerViewAnnotations[name] = value
+			}
+		}
+	}
+	return !reflect.DeepEqual(oldSyncerViewAnnotations, newSyncerViewAnnotations)
+}
+
+func SyncerViewsChanged(old, new metav1.Object) bool {
+	return FilteredSyncerViewsChanged(old, new, func(key string) bool {
+		return true
+	})
+}
+
+func SyncerViewChanged(old, new metav1.Object, syncTargetKey string) bool {
+	return FilteredSyncerViewsChanged(old, new, func(key string) bool {
+		return syncTargetKey == key
+	})
+}
+
+type Object interface {
+	metav1.Object
+	runtime.Object
+}
+
+func UpstreamViewChanged(old, new Object, equality func(old, new interface{}) bool) bool {
+	old = old.DeepCopyObject().(Object)
+	new = new.DeepCopyObject().(Object)
+
+	oldAnnotations := make(map[string]string)
+	for name, value := range old.GetAnnotations() {
+		if !strings.HasPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix) {
+			oldAnnotations[name] = value
+		}
+	}
+	old.SetAnnotations(oldAnnotations)
+
+	newAnnotations := make(map[string]string)
+	for name, value := range new.GetAnnotations() {
+		if strings.HasPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix) {
+			newAnnotations[name] = value
+		}
+	}
+	new.SetAnnotations(newAnnotations)
+
+	return !equality(old, new)
+}
+
+type SyncerViewManager[T Object] interface {
```

Comment:
managers have smell. What is this thing really?

Reply:
It is a object which can return the SyncerViews, synctarget-specifpc view of a resource for each sync target.
Sorry, doc is missing in the whole package. I'll add documentation here, and in the meantime will remove the corresponding commit from this PR.

Reply:
Remaned and documented